### PR TITLE
Extrusion

### PR DESCRIPTION
**Pull request by solis-lumine-vorago**
# Objective

- Adds a basic `Extrusion<T: Primitive2d>` shape, suggestion of #10572 

## Solution

- Adds `Measured2d` and `Measured3d` traits for getting the perimeter/area or area/volume of shapes. This allows implementing `.volume()` and `.area()` for all extrusions `Extrusion<T: Primitive2d + Measured2d>` within `bevy_math`
- All existing perimeter, area and volume implementations for primitves have been moved into implementations of `Measured2d` and `Measured3d`
- Shapes should be extruded along the Z-axis since an extrusion of depth `0.` should be equivalent in everything but name to the base shape

## Caviats

- I am not sure about the naming. `Extrusion<T>` could also be `Prism<T>` and the `MeasuredNd` could also be something like `MeasuredPrimitiveNd`. If you have any other suggestions, please fell free to share them :)

## Future work

This PR adds a basic `Extrusion` shape and does not implement a lot of things you might want it to. Some of the future possibilities include:
- [ ] bounding for extrusions
- [ ] making extrusions work with gizmos
- [ ] meshing

### CODE REVIEW

File: crates/bevy_math/src/primitives/dim3.rs [hunk 1]
```diff
@@ -915,6 +907,66 @@ impl Tetrahedron {
...
+    /// Get the volume of the tetrahedron.
+    #[inline(always)]
+    fn volume(&self) -> f32 {
+        self.signed_volume().abs()
+    }
+}
+
+/// A shape representiing an extruded `base_shape`.
+///
+/// The resulting volumes are prisms.
```
Comment (alice-i-cecile) [inline crates/bevy_math/src/primitives/dim3.rs:936]:
```suggestion
/// The resulting volumes are prisms.
/// For example, a triangle becomes a triangular prism, while a circle becomes a cylinder.
```

File: crates/bevy_math/src/primitives/dim3.rs [hunk 2]
```diff
@@ -915,6 +907,66 @@ impl Tetrahedron {
...
+    }
+
+    /// Get the volume of the tetrahedron.
+    #[inline(always)]
+    fn volume(&self) -> f32 {
+        self.signed_volume().abs()
+    }
+}
+
+/// A shape representiing an extruded `base_shape`.
```
Comment (alice-i-cecile) [inline crates/bevy_math/src/primitives/dim3.rs:934]:
```suggestion
/// A 3D shape representing an extruded 2D `base_shape`.
///
/// Extruding a shape effectively "thickens" a 2D shapes,
/// creating a shape with the same cross-section over the entire depth.
```

File: crates/bevy_math/src/primitives/mod.rs [hunk 3]
```diff
@@ -29,3 +29,20 @@ pub enum WindingOrder {
...
+
+    /// Get the area of the shape
+    fn area(&self) -> f32;
+}
+
+/// A trait for getting measurements of 3D shapes
+pub trait Measured3d {
+    /// Get the surface area of the shape
+    fn area(&self) -> f32;
+    /// Get the volume of the shape
```
Comment (mweatherley) [inline crates/bevy_math/src/primitives/mod.rs:46]: Giga nitpick: These have different line spacing

**Review by mweatherley — APPROVED** [review]
Looks good! My only misgiving with this is that the way that traits work makes some of the docs worse, but there's not really anything you can do about that.

File: crates/bevy_math/src/primitives/mod.rs (hunk 3 above)
Comment (solis-lumine-vorago) [reply crates/bevy_math/src/primitives/mod.rs:47]: haha, you're right, it's fixed now :)

**Comment by lynn-lumen** [comment]
> Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to [bevyengine/bevy-website#1312](https://github.com/bevyengine/bevy-website/issues/1312) if you'd like to help out.

I think it would make a lot of sense to merge either #13478 or #13346 first, because as it stands you really cannot do anything with extrusions

**Pull request merged**